File: challenges/week3.bag-of-tokens/ikasty.cpp
// Runtime: 4 ms, faster than 96.80% of C++ online submissions
// Memory Usage: 10.5 MB, less than 95.32% of C++ online submission
#include <algorithm>
#include <vector>

class Solution {
public:
    int bagOfTokensScore(std::vector<int>& tokens, int P) {
        std::sort(tokens.begin(), tokens.end());
        auto score = 0, max_score = 0;
        auto head = 0;
        auto tail = tokens.size();

        while (head != tail) {
            if (tokens[head] <= P) {
                P -= tokens[head++];
                score++;
            } else if (score > 0) {
                P += tokens[--tail];
                score--;
            }
            else break;

            if (score > max_score) max_score = score;
        }

        return max_score;
    }
};

File: challenges/week6.sort-the-matrix-diagonally/ikasty.py
# Runtime: 76 ms, faster than 96.00% of Python3 online submissions
# Memory Usage: 14.6 MB, less than 74.40% of Python3 online submissions
class Solution:
    def diagonalSort(self, mat: List[List[int]]) -> List[List[int]]:
        m, n = len(mat), len(mat[0])

        # row-wise sort
        for row in range(m):
            diagonal = sorted([ mat[i][j] for (i, j) in zip(range(row, m), range(n)) ])
            for (i, j, x) in zip(range(row, m), range(n), diagonal):
                mat[i][j] = x

        # column-wise sort
        for col in range(1, n):
            diagonal = sorted([ mat[i][j] for (i, j) in zip(range(m), range(col, n)) ])
            for (i, j, x) in zip(range(m), range(col, n), diagonal):
                mat[i][j] = x

        return mat